pallets/ddc-accounts/src/lib.rs

```diff
@@ -0,0 +1,603 @@
+		/// Number of eras that staked funds must remain bonded for.
+		#[pallet::constant]
+		type BondingDuration: Get<EraIndex>;
+		type TimeProvider: UnixTime;
+	}
+
+	/// Map from all locked "stash" accounts to the controller account.
+	#[pallet::storage]
+	#[pallet::getter(fn bonded)]
+	pub type Bonded<T: Config> = StorageMap<_, Twox64Concat, T::AccountId, T::AccountId>;
+
+	/// Map from all (unlocked) "controller" accounts to the debug regarding the staking.
+	#[pallet::storage]
+	#[pallet::getter(fn ledger)]
+	pub type Ledger<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AccountId, AccountsLedger<T::AccountId, BalanceOf<T>>>;
+
+	#[pallet::type_value]
+	pub fn DefaultBucketCount<T: Config>() -> u128 {
+		0_u128
+	}
+	#[pallet::storage]
+	#[pallet::getter(fn buckets_count)]
+	pub type BucketsCount<T: Config> =
+		StorageValue<Value = u128, QueryKind = ValueQuery, OnEmpty = DefaultBucketCount<T>>;
+
+	/// Map from all locked accounts and their buckets to the bucket structure
+	#[pallet::storage]
+	#[pallet::getter(fn buckets)]
+	pub type Buckets<T: Config> =
+		StorageMap<_, Twox64Concat, u128, Bucket<T::AccountId>, OptionQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// An account has bonded this amount. \[stash, amount\]
+		///
+		/// NOTE: This event is only emitted when funds are bonded via a dispatchable. Notably,
+		/// it will not be emitted for staking rewards when they are added to stake.
+		Deposited(T::AccountId, BalanceOf<T>),
+		/// An account has unbonded this amount. \[stash, amount\]
+		Unbonded(T::AccountId, BalanceOf<T>),
+		/// An account has called `withdraw_unbonded` and removed unbonding chunks worth `Balance`
+		/// from the unlocking queue. \[stash, amount\]
+		Withdrawn(T::AccountId, BalanceOf<T>),
+		/// Total amount charged from all accounts to pay CDN nodes
+		Charged(BalanceOf<T>),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Not a controller account.
+		NotController,
+		/// Not a stash account.
+		NotStash,
+		/// Stash is already bonded.
+		AlreadyBonded,
+		/// Controller is already paired.
+		AlreadyPaired,
+		/// Cannot deposit dust
+		InsufficientDeposit,
+		/// Can not schedule more unlock chunks.
+		NoMoreChunks,
+		/// Internal state has become somehow corrupted and the operation cannot continue.
+		BadState,
+	}
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig;
+
+	#[cfg(feature = "std")]
+	impl Default for GenesisConfig {
+		fn default() -> Self {
+			Self
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig {
+		fn build(&self) {
+			let account_id = <Pallet<T>>::account_id();
+			let min = T::Currency::minimum_balance();
+			if T::Currency::free_balance(&account_id) < min {
+				let _ = T::Currency::make_free_balance_be(&account_id, min);
+			}
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(10_000)]
+		pub fn charge_payments(
```

Comment:
Could be an internal function in case it does not require state transition by the OCW. Otherwise, it should be extrinsic due to the inability of the OCW to do a state transition without sending a transaction.